### PR TITLE
Add --watch mode to `flutter test`

### PR DESCRIPTION
**Pull request by gamebox**
When flag is provided watch application directory for file changes and re-run tests on change, only incrementally compiling changed files. This provides a significant speedup on incremental edit/test cycle.

### CODE REVIEW

**Comment by gamebox** [comment]
@jonahwilliams Updated with a much more readable series of commits.

**Comment by jonahwilliams** [comment]
I haven't forgotten about this! Taking a longer look at it today.

**Comment by gamebox** [comment]
The test failures are expected due to the facts laid out in the commit message of the commit just added

**Comment by jonahwilliams** [comment]
Refactoring is fine but not necessary to this PR.

I pulled your gamebox/flutter and pointed test at gamebox/test with dependency overrides. `flutter test --watch` ran correctly the first time, but it did not detect any changes to the test directory or lib. I also could not exit the process without CTRL+C.

What is the intended workflow here? am I watching everything, a specific test file, or a library file?. I can see different cases where a watch command is useful depending on what I am working on:

- I am adjusting a library file and I want to rerun one test, multiple tests, multiple test files on change
- I am adjusting a test file and I want to rerun it on change.

Have you considered any other approaches? For example, I _could_ wire up either command myself with some shell commands, but it would be too slow. Thus it seems like caching the kernel files to disk might also speed up incremental runs. Of course, you would have to be more careful with invalidation!

**Comment by apaatsio** [comment]
I'll just mention https://github.com/flutter/flutter/issues/4719 here since it's relevant.

**Comment by gamebox** [comment]
Thanks @apaatsio , I'll check in with @Hixie about the operational semantics of this feature.

**Comment by gamebox** [comment]
@jonahwilliams Just an update, my change dart-lang/test was approved.  Once it's merged, @grouma will cut a new release, and then I can update this branch to use it

**Comment by gamebox** [comment]
Alright @jonahwilliams , this includes the dart-lang/test updates, so this should reflect generally what I would like to release as a WIP feature in the tool.  The semantics have been changed from rerun-on-save, to the same as hot reload('r' to rerun, 'q' to quit).

File: packages/flutter_tools/pubspec.lock [hunk 1]
```diff
@@ -0,0 +1,502 @@
+# Generated by pub
```
Comment (jonahwilliams) [inline packages/flutter_tools/pubspec.lock]: Make sure to revert this file

**Comment by jonahwilliams** [comment]
Got everything working locally, this is pretty neato!

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 2]
```diff
@@ -68,19 +72,51 @@ Future<int> runTests(
...
-    machine: machine,
-    startPaused: startPaused,
-    serverType: serverType,
-    precompiledDillPath: precompiledDillPath,
-    precompiledDillFiles: precompiledDillFiles,
-    trackWidgetCreation: trackWidgetCreation,
-    updateGoldens: updateGoldens,
-    projectRootDirectory: fs.currentDirectory.uri,
-  );
+  final Uri projectRootDirectory = fs.currentDirectory.uri;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart]: Consider removing some of the empty lines here

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 3]
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
 
   // Call package:test's main method in the appropriate directory.
   final Directory saved = fs.currentDirectory;
+  
+  final Function printReloadMessage = () {
+    printStatus(
+      "Press 'r' to hot reload your tests, 'q' to quit",
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart]: Instead of hot reload call this rerunning, it is more descriptive.

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 4]
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
...
+        if (event.path.endsWith('.dart')) {
+          invalidatedFiles.add(event.path);
+        }
+      });
+
+      terminal.singleCharMode = true;
+      terminal.onCharInput.asBroadcastStream().listen((String char) async {
+        switch(char) {
+          case 'r':
+            stdout.writeln(terminal.clearScreen());
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart]: I really don't like clearing the terminal here, I lose the context of my previous test run/failure.

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 5]
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
...
+            printStatus('Recompiling test files...\n', emphasis: true, color: TerminalColor.blue);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart]: I would use a Status instead:
```
Status status = logger.startProgress('Recompiling test files...', timeout: null);
...
status.stop();

```

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 6]
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
...
+            await compileTestFiles();
+            await test.runTests(testArgs);
+
+            printReloadMessage();
+            break;
+          case 'q':
+            completer.complete();
+        }
+      }, onDone: completer.complete);
+
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart]: Consider removing some of these empty lines

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 7]
```diff
@@ -89,14 +125,65 @@ Future<int> runTests(
...
+      printReloadMessage();
+
+      await completer.future;
+      stdout.writeln(terminal.clearScreen());
+    }
+
+    try {
+      test.completeShutdown();
+    } on StateError catch (e) {
+      printTrace(e.toString());
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/test/runner.dart]: printError ?

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 8]
```diff
@@ -124,6 +132,10 @@ class TestCommand extends FlutterCommand {
 
     Directory workDir;
     if (files.isEmpty) {
+      if (argResults['watch']) {
+        throwToolExit(
+          'When using --watch, you must specify one or more test file(s) to run.');
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart]: Presumably the user could manually specify every single file right? Might as well make that supported fully

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 9]
```diff
@@ -167,14 +182,18 @@ class TestCommand extends FlutterCommand {
...
+      watchTests: watchTests,
       trackWidgetCreation: argResults['track-widget-creation'],
       updateGoldens: argResults['update-goldens'],
       concurrency: jobs,
     );
 
     if (collector != null) {
-      if (!await collector.collectCoverageData(
-          argResults['coverage-path'], mergeCoverageData: argResults['merge-coverage']))
+      final bool coverageDataCollected = await collector.collectCoverageData(
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart]: Revert this change

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 10]
```diff
@@ -139,7 +151,8 @@ class TestCommand extends FlutterCommand {
     }
 
     CoverageCollector collector;
-    if (argResults['coverage'] || argResults['merge-coverage']) {
+    if (argResults['coverage'] ||
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart]: Revert this change

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 11]
```diff
@@ -75,6 +75,14 @@ class TestCommand extends FlutterCommand {
         help: 'Whether matchesGoldenFile() calls within your test methods should '
               'update the golden files rather than test for an existing match.',
       )
+      ..addFlag('watch',
+        defaultsTo: false,
+        negatable: false,
+        hide: true, // TODO(gamebox): Remove when UX is finalized
+        help: 'After initial run, continue watching application directory for '
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart]: "similar semantics to hot reload" - is that actually true? do constants not change between test runs now?

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 12]
```diff
@@ -75,6 +75,14 @@ class TestCommand extends FlutterCommand {
         help: 'Whether matchesGoldenFile() calls within your test methods should '
               'update the golden files rather than test for an existing match.',
       )
+      ..addFlag('watch',
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/test.dart]: Even though there is some file watching going on, I don't think that is the defining feature of this. Maybe "interactive?" I'm not quite sure.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 8 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: The philosophy here is to make the main use case easy, and what we just allow to be done, but with some difficulty.  A sensible alternative would be to restrict to a single test file.  I don't have any _strong_ opinions either way, but we are solving for taking a single - or a small number of - test(s) from red to green.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 11 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Sorry, I'm not sure what you mean by "do constants not change between test runs now?"  We can discuss offline.  By "similar semantics", I meant we only recompile the changed files and then rerun the test code - not to mention that the rerunning happens by pressing 'r' in the terminal/stdin.  Which is why I did not say "the exact same semantics".

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 12 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: I'm open to either, I have it hidden so that we can feel free to change this in the near future.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/test.dart]: I agree in principle, but there is an edge case where the behavior is painful enough that we may as well allow it: the project as one or few tests files

**Comment by gamebox** [comment]
Will address the rest of the comments in the morning.  Thanks for the thorough review.  Most of my issues remaining have to do with updating the dependencies.  Will discuss offline.

File: packages/flutter_tools/lib/src/commands/test.dart [hunk 13]
```diff
@@ -124,6 +132,10 @@ class TestCommand extends FlutterCommand {
 
     Directory workDir;
     if (files.isEmpty) {
+      if (argResults['watch']) {
```
[the remark this answers is not in the document]
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Done

File: packages/flutter_tools/lib/src/test/bootstrap.dart [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+      }
+      directory = directory.parent;
+    }
+    return _generateTestBootstrap(
+      testUrl: testUrl,
+      testConfigFile: testConfigFile,
+      host: host,
+      updateGoldens: updateGoldens,
+    );
+  }
```
[the remark this answers is not in the document]
Comment (gamebox) [reply packages/flutter_tools/lib/src/test/bootstrap.dart]: Done

File: packages/flutter_tools/lib/src/test/bootstrap.dart [hunk 15]
```diff
@@ -0,0 +1,181 @@
+import 'package:meta/meta.dart';
```
[the remark this answers is not in the document]
Comment (gamebox) [reply packages/flutter_tools/lib/src/test/bootstrap.dart]: Done

File: packages/flutter_tools/lib/src/test/runner.dart (hunk 2 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/test/runner.dart]: Done

File: packages/flutter_tools/lib/src/test/runner.dart (hunk 3 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/test/runner.dart]: Done

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 9 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Done

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 10 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Done

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 9 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Done

File: packages/flutter_tools/lib/src/test/runner.dart (hunk 4 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/test/runner.dart]: Removed, I thought it was cleaner to see what the current test state is while running.

File: packages/flutter_tools/lib/src/test/runner.dart (hunk 5 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/test/runner.dart]: Didn't see this.  This is awesome.  Done

File: packages/flutter_tools/lib/src/test/runner.dart (hunk 7 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/test/runner.dart]: Of course, Done.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 8 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Agreed offline to remove this stricture

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 12 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Leaving as is until someone has a better name for the flag.  Since the files are in fact being watched, I think the current name makes sense.

File: packages/flutter_tools/lib/src/commands/test.dart (hunk 11 above)
Comment (gamebox) [reply packages/flutter_tools/lib/src/commands/test.dart]: Changed the language to reduce confusion.

**Comment by gamebox** [comment]
That last "queued" test has actually completed and passed.  Don't know if that's something wrong on Cirrus or GH's side.

**Comment by gamebox** [comment]
Closing in favor of (#27192)
